metacat-main/src/main/java/com/netflix/metacat/main/manager/ConnectorManager.java

```diff
@@ -103,31 +116,91 @@ public void addPlugin(final ConnectorPlugin connectorPlugin) {
-                                              final ConnectorContext connectorContext) {
+    public synchronized void createConnection(final ConnectorContext connectorContext) {
         Preconditions.checkState(!stopped.get(), "ConnectorManager is stopped");
+        final String connectorType = connectorContext.getType();
+        final String catalogName = connectorContext.getCatalogName();
         final ConnectorPlugin connectorPlugin = plugins.get(connectorType);
         if (connectorPlugin != null) {
-            Preconditions
-                .checkState(!connectorFactories.containsKey(catalogName), "A connector %s already exists", catalogName);
-            final ConnectorFactory connectorFactory =
-                connectorPlugin.create(catalogName, connectorContext);
-            connectorFactories.put(catalogName, connectorFactory);
-
             final MetacatCatalogConfig catalogConfig =
-                MetacatCatalogConfig.createFromMapAndRemoveProperties(connectorType,
+                MetacatCatalogConfig.createFromMapAndRemoveProperties(connectorType, catalogName,
                     connectorContext.getConfiguration());
-            catalogs.put(catalogName, catalogConfig);
+            final List<String> databaseNames = catalogConfig.getSchemaWhitelist();
+            if (databaseNames.isEmpty()) {
+                Preconditions.checkState(!catalogs.contains(catalogName, EMPTY_STRING),
+                    "A catalog with name %s already exists", catalogName);
+            } else {
+                databaseNames.forEach(databaseName -> {
+                    Preconditions.checkState(!catalogs.contains(catalogName, databaseName),
+                        "A catalog with name %s for database %s already exists", catalogName, databaseName);
+                });
+            }
+            catalogConfigs.add(catalogConfig);
+            final ConnectorFactory connectorFactory = connectorPlugin.create(connectorContext);
+            try {
+                databaseServices.add(connectorFactory.getDatabaseService());
+            } catch (UnsupportedOperationException e) {
+                log.debug("Catalog {} doesn't support getDatabaseService. Ignoring.", catalogName);
+            }
+            try {
+                tableServices.add(connectorFactory.getTableService());
+            } catch (UnsupportedOperationException e) {
+                log.debug("Catalog {} doesn't support getTableService. Ignoring.", catalogName);
+            }
+            try {
+                partitionServices.add(connectorFactory.getPartitionService());
+            } catch (UnsupportedOperationException e) {
+                log.debug("Catalog {} doesn't support getPartitionService. Ignoring.", catalogName);
+            }
+            final CatalogHolder catalogHolder = new CatalogHolder(catalogConfig, connectorFactory);
+            if (databaseNames.isEmpty()) {
+                catalogs.put(catalogName, EMPTY_STRING, catalogHolder);
```

Comment:
Maybe this should be moved to the top along with the first
`if (databaseNames.isEmpty()) {`

Reply:
this is empty represents everything ( as default) or a real empty database?  Are we use empty string as a placeholder for the catalogs having no database?